accepted/wasmtime-baseline-compilation.md

```diff
@@ -0,0 +1,259 @@
+# Baseline Compilation in Wasmtime
+
+Authors: Saúl Cabrera (@saulecabrera); Chris Fallin (@cfallin)
+
+## Summary
+
+This RFC proposes the addition of a new WebAssembly (Wasm) compiler to Wasmtime:
+a single-pass or “baseline” compiler. A baseline compiler works to improve
+overall compilation performance, yielding faster startup times, at the cost of
+less optimized code.
+
+
+## Motivation
+
+Wasmtime currently uses Cranelift by default, which is an optimizing compiler.
+Cranelift performs code optimizations at the expense of slower compilation
+times. This makes Just-In-Time (JIT) compilation of Wasm unsuitable for cases
+where higher compilation performance is desired (e.g. short-lived trivial
+programs, cases in which startup time is more critical than runtime
+performance). 
+
+The introduction of a baseline compiler is a first step towards: (i) faster
+compilation and startup times (ii) enabling a tiered compilation model in
+Wasmtime, similar to what is present in Wasm engines in web browsers. This RFC
+**does not** account for tiered compilation, it only accounts for the
+introduction of a baseline compiler.
+
+**Approximate** [measurements](https://github.com/Shopify/wasm-bench) taken on
+top of a subset of the [Sightglass benchmark
+suite](https://github.com/bytecodealliance/sightglass/tree/main/benchmarks)
+– using different optimizing and baseline compilers (Cranelift from Wasmtime,
+Liftoff from V8 and RabaldrMonkey from SpiderMonkey) – show that a baseline
+compiler on average yields 15x to 20x faster compilation while producing code
+that is on average 1.1x to 1.5x slower than the one produced by an optimizing
+compiler. These measurements align on average with other measurements observed
+when comparing interpretation and compilation for WebAssembly[^1].
```

Comment:
As a source of inspiration for design (edit: now I see the design principles are very close to what's in this blog post), as well as numbers showing the speed of a baseline compiler, I recommend this great post by Andy Wingo: https://wingolog.org/archives/2020/03/25/firefoxs-low-latency-webassembly-compiler

Reply:
Indeed a great post! I've used it as a source of inspiration and as a guide to navigate SpiderMonkey's code.